cmd/autoretrieve/main.go

```diff
@@ -0,0 +1,772 @@
+		}
+
+		api, closer, err := lcli.GetGatewayAPI(cctx)
+		if err != nil {
+			return err
+		}
+		defer closer()
+
+		nd, err := newAutoRetrieveNode(cctx.Context, Config{
+			retrievalTimeout: time.Second * time.Duration(cctx.Int("timeout")),
+			dataDir:          dataDir,
+			listenAddrs:      []multiaddr.Multiaddr{multiaddr.StringCast("/ip4/0.0.0.0/tcp/6746")},
+			minerBlacklist:   minerBlacklist,
+		}, api)
+		if err != nil {
+			return err
+		}
+
+		logger.Infof("P2P Address: %v", nd.host.Addrs())
+		logger.Infof("P2P ID: %v", nd.host.ID())
+
+		<-cctx.Context.Done()
+
+		return nil
+	}
+
+	// Clean exit on first interrupt, hard stop on second
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		ch := make(chan os.Signal, 1)
+		signal.Notify(ch, os.Interrupt)
+
+		<-ch
+
+		signal.Ignore()
+		fmt.Printf("Interrupt detected, gracefully exiting... (interrupt again to force termination)\n")
+		cancel()
+	}()
+
+	err := app.RunContext(ctx, os.Args)
+	if err != nil {
+		log.Fatalln(err)
+	}
+}
+
+const datastoreSubdir = "datastore"
+const blockstoreSubdir = "blockstore"
+const walletSubdir = "wallet"
+const minerBlacklistFilename = "blacklist.txt"
+
+func readMinerBlacklist(dataDir string) ([]address.Address, error) {
+	bytes, err := os.ReadFile(filepath.Join(dataDir, minerBlacklistFilename))
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return nil, nil
+		}
+
+		return nil, err
+	}
+
+	strs := strings.Split(string(bytes), "\n")
+
+	var miners []address.Address
+	for lineNum, str := range strs {
+		str = strings.TrimSpace(str)
+
+		if str == "" {
+			continue
+		}
+
+		miner, err := address.NewFromString(str)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse miner at line %v: %v", lineNum, err)
+		}
+
+		miners = append(miners, miner)
+	}
+
+	return miners, nil
+}
+
+type Config struct {
+	retrievalTimeout time.Duration
+	dataDir          string
+	listenAddrs      []multiaddr.Multiaddr
+	minerBlacklist   map[address.Address]bool
+}
+
+type autoRetrieveNode struct {
+	datastore  datastore.Batching
+	blockstore *senderBlockstore
+	wallet     *wallet.LocalWallet // If nil, only free retrievals will be attempted
+	fc         *filclient.FilClient
+	host       host.Host
+	bsnet      bsnet.BitSwapNetwork
+}
+
+func newAutoRetrieveNode(ctx context.Context, config Config, api api.Gateway) (*autoRetrieveNode, error) {
+	node := &autoRetrieveNode{}
+
+	// Datastore
+	{
+		datastore, err := leveldb.NewDatastore(filepath.Join(config.dataDir, datastoreSubdir), nil)
+		if err != nil {
+			return nil, err
+		}
+
+		node.datastore = datastore
+	}
+
+	// Blockstore
+	{
+		parseShardFunc, err := flatfs.ParseShardFunc("/repo/flatfs/shard/v1/next-to-last/3")
+		if err != nil {
+			return nil, err
+		}
+
+		blockstoreDatastore, err := flatfs.CreateOrOpen(filepath.Join(config.dataDir, blockstoreSubdir), parseShardFunc, false)
+		if err != nil {
+			return nil, err
+		}
+
+		node.blockstore = &senderBlockstore{
```

Comment:
its probably worth throwing some metrics wrapping in here, both above and below the auto-retrieve bits so we can measure them more easily

Reply:
see "measure blockstore" elsewhere